script/replace_macros.sh

```diff
@@ -0,0 +1,117 @@
+#!/bin/bash
+#
+#  Copyright (c) 2019, The OpenThread Authors.
+#  All rights reserved.
+#
+#  Redistribution and use in source and binary forms, with or without
+#  modification, are permitted provided that the following conditions are met:
+#  1. Redistributions of source code must retain the above copyright
+#     notice, this list of conditions and the following disclaimer.
+#  2. Redistributions in binary form must reproduce the above copyright
+#     notice, this list of conditions and the following disclaimer in the
+#     documentation and/or other materials provided with the distribution.
+#  3. Neither the name of the copyright holder nor the
+#     names of its contributors may be used to endorse or promote products
+#     derived from this software without specific prior written permission.
+#
+#  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+#  AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+#  IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+#  ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+#  LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+#  CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+#  SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+#  INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+#  CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+#  ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+#  POSSIBILITY OF SUCH DAMAGE.
+#
+#  Usage:
+#  replace_macros.sh OPENTHREAD_CONFIG_HEADER_FILE
+
+sed -i \
```

Comment:
While I agree having such scripts is very helpful in transitioning projects from older config definitions to new model, I am not sure if it makes sense to include this in OpenThread GitHub master branch. OT master does not support any of the older config definitions. I think including such a script can be confusing to a new developer/user of OT exploring the code.

In the history of development of OpenThread, we have had changes to the public OT APIs. Two larger ones that I recall:

- [PR #1764](https://github.com/openthread/openthread/pull/1764) where `otError` enumeration names  were changed from `kThreadError<ErrorInCamelCase>` model to `OT_ERROR_<ErrorInUpperCase>` model (the type itself was renamed `ThreadError` -> `otError`).
- [PR #1173 ](https://github.com/openthread/openthread/pull/1173) where many of public APIs and definitions were renamed to follow a consistent coding/naming convention. 

In such cases, we used scripts and tools to update all internal projects and convert the old naming model to new one, but we did not merge/include the scripts in the OpenThread master (github master branch). 

Thoughts?

Reply:
Maybe we can put this script on gist and provide a link in openthread-core-config-check.h?

Reply:
@abtink I think this tool is mainly for OpenThread users, who might still use a very old version of OpenThread(I used to find this fact in some questions raised in openthread-users group).  And once they want to upgrade to the latest OpenThread, they will find it difficult. That's why I think it makes sense to provide and maintain a script to migrate old code to latest code.

Although we don't support any of the older config definitions, we still provided `openthread-config-default-check.h` to check those old definitions are not defined. That's helpful to report a problem. But If I am an OpenThread user, I would wish to have a tool to help me solve the problem if possible. I was just asked to help migrate https://github.com/openthread/silk, I find this tool is very helpful.

So I tend to maintain this script, and every time we make a change, we should update this script to migrate the code. This also helps us to do GitHub sync. Thoughts?

Reply:
@bukepo, @gjc13 I 100% agree that such script would be useful. Only considering where to save/store such scripts (master or ...?)
On some internal project/repos, we used a `scripts` branch that included useful scripts for that repo. I think `wpantund` adopted same model [(scripts branch)](https://github.com/openthread/wpantund/tree/scripts).
I also like the idea of saving it on a gist.

Reply:
@jwhui Could you shed some lights on this issue?